File: h/反转链表.java
/*
输入一个链表，反转链表后，输出新链表的表头
*/
/*
public class ListNode {
    int val;
    ListNode next = null;
 
    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode ReverseList(ListNode head) {
        // 空表
        if (head == null) {
            return null;
        }
         
        // 只有一个节点
        if (head.next == null) {
            return head;
        }
         
        // 心得与头节点
        ListNode newHead = null;
        ListNode preNode = null;
        ListNode curNode = head;
        ListNode nextNode = null;
         
        while (curNode != null) {
            nextNode = curNode.next;
            // 到末尾
            if (nextNode == null) {
                newHead = curNode;
            }
            curNode.next = preNode;
            preNode = curNode;
            curNode = nextNode;
          
        }
        return newHead;
         
    }
}

File: z/反转链表.cc
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
            val(x), next(NULL) {
    }
};*/
class Solution {
 public:
  ListNode* ReverseList(ListNode* pHead) {
    if (pHead == NULL) {
      return NULL;
    }
    ListNode* prev = NULL;
    while (pHead != NULL) {
      ListNode* tmp = pHead->next;
      pHead->next = prev;
      prev = pHead;
      pHead = tmp;
    }
    return prev;
  }
};
